### ISSUE TITLE
Investigate / implement Proxy-Tuning (Liu et al. 2024) decode-time logit arithmetic

### ISSUE THREAD

**Issue by dpalmqvist** [motsognirr/olmlx#523]
## Summary

Investigate adding **Proxy-Tuning** (Liu et al. 2024, *Tuning Language Models by Proxy*) to olmlx. It steers a large base model **M** at decode time — without touching its weights — using a small tuned **expert M⁺** and small untuned **anti-expert M⁻** (M⁺ is M⁻ fine-tuned). Each decode step:

```
logits_final = logits_M + α · (logits_M⁺ − logits_M⁻)
```

then sample normally. The learned "tuning delta" from the cheap small models is transplanted onto the big model. Hard constraint: **all three models must share one exact tokenizer/vocabulary**, and all three run every decode step.

## Where it fits

Structurally it resembles olmlx's speculative decoding (multiple models, one decode loop) with one key difference: speculative decoding is *exactness-preserving*; proxy-tuning *deliberately alters* the output distribution, so it cannot reuse the draft→verify→accept logic. It is a **sibling decode mode**, not a speculative strategy.

### Recommended approach (Option B): a `SpecDecoderBase` subclass

Build on the speculative plumbing, swapping verify for logit arithmetic:

- Implement `_prefill_impl()` / `_step_impl()` / `_reset_state()` (`engine/spec_decoder_base.py:173–430`), holding three `(model, cache)` pairs instead of target+draft. Gets exception-safety try/reset + lifecycle for free.
- Reuse the prefill choreography from `_drive_spec_prefill` (`engine/speculative.py:399–470`): N models prefilled in lockstep, chunked at 2048 tokens with `mx.eval`/`mx.clear_cache()` between chunks, **no `mx.stream` wrapper** so it lands on `default_stream`.
- Replace `verify_draft_greedy` (`engine/speculative.py:125–157`) with the combiner: one forward per model, then `base + α·(expert − antiexpert)` before sampling.
- Each model gets an independent KV cache with **no trimming interaction** (simpler than classic speculative, which trims to the accepted prefix).

### Alternative (Option A, not recommended): logits processor

A grammar-style processor (`engine/logits_processors.py:74–130`, signature `__call__(tokens, logits)` per `engine/grammar.py:244–328`) is minimal but rides the **non-speculative** path on `generation_stream` and forces awkward KV-cache smuggling into a closure (the processor receives full token history each call). Worse fit for 3 concurrent models.

## Critical constraints / gotchas

1. **Vocabulary identity is stricter than what exists.** `speculative_loaders._check_vocab_match` (`engine/speculative_loaders.py:99–115`) only compares the integer `vocab_size`; two models can match on size but differ in token mapping — fatal for logit arithmetic. Need a stronger guard (tokenizer hash / vocab equality). The M⁺/M⁻ pair share vocab trivially; the real question is base ↔ pair (the family question below).
2. **Metal stream hazard.** Three models means keep all forwards on one stream (`default_stream`, as the spec path does) or risk the GatedDeltaNet/MoE expert-routing corruption documented in CLAUDE.md. The classic decoder already detects hybrids and installs a single shared `GDNStateCapture` (`engine/speculative.py:543–593`), raising if two models use different GDN classes.
3. **Loading & pinning.** `ModelManager.ensure_loaded()` + `acquire_ref()`/`active_refs` (`engine/model_manager.py:609–617, 939, 1052–1065`) lets all three coexist and prevents mid-inference eviction — requires running with `max_loaded_models ≥ 3` (same note as the panel coordinator).
4. **Compatibility.** Grammar works *after* the combination (it just masks). Speculative decoding cannot run simultaneously — both want to own the loop; gate mutually exclusive (grammar already disables speculative at `engine/inference.py:3338–3348`).
5. **Cost.** No amortization: per token ≈ `base_forward + 2 × small_forward`. Worth it only when expert/anti-expert are much smaller than base.

## Candidate model families

Binding requirement: a **(base, expert, anti-expert) triple under one tokenizer** — i.e. families shipping a wide size ladder with a shared vocab. Cross-family is impossible (different tokenizers).

| Family | Fit | Notes |
|---|---|---|
| **Qwen2.5 / Qwen3 (dense)** | ★ Strongest | One tokenizer (vocab 151936) across 0.5B→72B; ideal 0.6–1.5B pair steering 14–72B. |
| **Llama 3 / 3.1** | ★ Strong | Shared tokenizer across 8B/70B/405B; the paper's original setting (Llama-2). |
| **Gemma 2** | ✓ Good | Shared 256k vocab across 2B/9B/27B; watch memory from large vocab. |
| **Mistral / Mixtral** | ✓ Good | Shared tokenizer; fewer small/large pairings than Qwen/Llama. |
| **Gemma 3** | △ Caveats | Pure-`RotatingKVCache` ⇒ single-chunk-prefill + non-trimmable cache complicate 3-model KV. |
| **Qwen3.5 / Qwen3-Next (hybrid)** | △ Caveats | Linear-attention hybrids trigger the Metal stream hazard hardest; must stay on `default_stream`, no cross-request cache reuse. |
| **gpt-oss** | ✗ Weak | Only 20B/120B (no cheap small pair) + pure rotating cache. |
| **DeepSeek V2/V3** | ✗ Weak | Vocab shared but no small expert/anti-expert sizes available. |

**Recommendation:** target **Qwen2.5/Qwen3 dense** and **Llama 3.x** first; Gemma 2 and Mistral secondary.

## Proposed scope (incremental)

- [ ] Vocab-identity guard (tokenizer/vocab equality, beyond `vocab_size`).
- [ ] `ProxyTuningDecoder(SpecDecoderBase)` with 3 models + independent KV caches, all on `default_stream`.
- [ ] CLI subcommand + `config.py` knobs (base/expert/anti-expert paths, `α`), mirroring `--speculative-*`.
- [ ] Mutual-exclusion gating vs speculative; grammar applied after combination.
- [ ] TDD per CLAUDE.md: failing tests first; validate on a Qwen3 base + small expert/anti-expert pair.

_Investigation tracked from session; design grounded in `engine/spec_decoder_base.py`, `engine/speculative.py`, `engine/speculative_loaders.py`, `engine/model_manager.py`, and `engine/logits_processors.py`._

---

## Validation (code-audited 2026-06-14)

Claims above checked against the live tree (`olmlx/olmlx/engine/…` — body paths drop the package prefix; some cited line ranges are off, noted below). The Option B design holds; corrections and two underweighted risks follow.

### ✅ Confirmed
- **`SpecDecoderBase` subclassing fits.** Three abstract hooks `_prefill_impl` / `_step_impl` / `_reset_state` at `spec_decoder_base.py:351–370`; try/reset exception safety on `prefill()`/`step()` + `__del__→close→reset` lifecycle are free. Model/cache storage is **not** hardcoded to 2 — the base only names `_target`/`_draft` for teardown; a proxy decoder that leaves `_patched`/`_bound` False (defaults) skips those branches and stores 3 models under any names.
- **Full logit control, but greedy-only.** Every spec decoder is greedy `argmax` with **no sampler/temperature/top-p hook** anywhere in the spec path. A subclass can compute `base + α·(expert − antiexpert)` and argmax freely — but there is **no temperature sampling to reuse**; stochastic sampling must be implemented in the subclass if needed. (`verify_draft_greedy` lives at `spec_decoder_base.py:125–157`, not `speculative.py`; it consumes already-argmax'd choices, so the combiner needs the raw per-model logits.)
- **`_drive_spec_prefill` (`speculative.py:399–470`) is already N-generic** — takes `lanes: list[(model, cache)]`, 2048-token chunks, `mx.eval`/`mx.clear_cache()` between, no `mx.stream` ⇒ `default_stream`. Extends to 3 lanes for free.
- **Trim-avoidance is real.** Classic trims caches to the accepted prefix (`speculative.py:887–923`) only because draft/target advance by different amounts. Proxy-tuning advances every cache over the same committed sequence ⇒ no trim/rollback machinery. Genuine simplification.

### ⚠️ Corrections
1. **Pinning claim (gotcha #3) is inaccurate.** The existing speculative loader does **not** pin two `ModelManager` entries via `acquire_ref`/`active_refs`. Only the **target** is a tracked `LoadedModel`; the **draft is loaded inline via mlx-lm and held by reference on the decoder** (`speculative_loaders.py:573–641`), never ref-counted, never in `_loaded`. So the 3 models need a deliberate choice:
   - **(A) draft-style inline load** — expert/anti-expert invisible to LRU, but bypass the `generate_chat` Metal-stream/inference-lock handling.
   - **(B) panel-style `ensure_loaded(pin=True)`** — proper eviction-pinning via `active_refs` (LRU skips `active_refs > 0`, `model_manager.py:923–948`), but needs `max_loaded_models ≥ 3` (default is **1**, `config.py:101`). **Recommended** — keeps the stream/lock handling intact.
2. **Gating + dispatch are duplicated at two sites, not one.** Grammar↔speculative mutual exclusion exists in **both** the streaming path (`inference.py:3338–3348`) and the non-streaming path (`3934–3946`), with the new-mode branch selected at `3359` / `3975` respectively. A proxy-tuning mode must touch **both**.
3. **Vocab-guard plumbing gap.** `_check_vocab_match` (`speculative_loaders.py:98–115`) is confirmed `vocab_size`-only **and soft-fails to a warning** when `vocab_size` is absent. It currently receives **model objects, not tokenizers**. A real guard via `tokenizer.get_vocab()` equality/hash is feasible (the `TokenizerWrapper` forwards `get_vocab()`), but the tokenizer objects must be plumbed in — the draft loader currently **discards** its `_draft_tokenizer`.

### 🔴 Two risks underweighted above
1. **Hybrid/GDN models are a hard wall, not a "caveat."** The shared `GDNStateCapture` (`speculative.py:543–593`) **raises `NotImplementedError` if two models use different GDN classes** (patch lock is per-class). A base + expert + anti-expert that don't all share one GDN subclass cannot run. This downgrades the △ hybrid families (Qwen3.5/Next, Gemma 3) from "awkward" to **blocked** for v1 — reinforcing dense Qwen/Llama first (non-hybrid ⇒ no GDN capture at all).
2. **The decode loop is genuinely 2-model.** Only `_drive_spec_prefill` is N-ready; `_step_linear`/`_step_tree`, the cache/buffer fields, and the snapshot tuple `(target_snap, draft_snap)` are hardcoded pairs. `ProxyTuningDecoder` should therefore be a **fresh sibling subclass** with its own 3-model step loop — not a generalization of `SpeculativeDecoder`.

### Tightened scope
- [ ] Vocab-identity guard via `get_vocab()` equality/hash across all three tokenizers; plumb tokenizer objects (stop discarding `_draft_tokenizer`-style handles); keep it a **hard** error (no soft warn-and-pass).
- [ ] `ProxyTuningDecoder(SpecDecoderBase)` as a **fresh** subclass: 3 model/cache attrs, no layer hooks / no draft bind / **no GDN capture** / no trimming; greedy combine + argmax (add sampling only if required).
- [ ] Reuse `_drive_spec_prefill` with 3 lanes on `default_stream`.
- [ ] Loading via **(B)** `ensure_loaded(pin=True)` for all three + a `max_loaded_models ≥ 3` preflight check.
- [ ] Mutual-exclusion gate + dispatch branch mirrored at **both** `inference.py` sites (streaming 3338/3359, non-streaming 3934/3975).
- [ ] CLI subcommand + `config.py` knobs (base/expert/anti-expert paths, `α`), mirroring `--speculative-*`.
- [ ] **v1 target: dense Qwen2.5/Qwen3 + Llama 3.x only** (non-hybrid). Hybrid/GDN and pure-RotatingKVCache families deferred until the single-GDN-class wall and single-chunk-prefill are addressed.
- [ ] TDD per CLAUDE.md: failing tests first; validate on a Qwen3 dense base + small expert/anti-expert pair.

**Comment by dpalmqvist** [comment]
Implemented and merged in #525 (merge commit 8204a5e).

**Shipped (v1):**
- \`ProxyTuningDecoder(SpecDecoderBase)\` — decode-time \`base + α·(expert − antiexpert)\`, registered as \`speculative_strategy="proxy_tuning"\` to reuse the speculative lifecycle/stream/dispatch/grammar-gate wiring (no \`inference.py\` changes).
- Expert/anti-expert **loaded inline by the loader** (held by the decoder, not in \`ModelManager._loaded\`) → no \`max_loaded_models\` bump / \`ensure_loaded\` pinning needed. This superseded the issue's original "strategy B" recommendation; rationale in the code-audited investigation above.
- Stronger \`check_vocab_identity\` (token-mapping equality) on top of the size-only \`_check_vocab_match\`; flash-MoE mutual-exclusion; CLI startup-audit + \`--speculative-strategy\` support; eager \`close()\` to free expert/anti-expert weights on unload; metrics label.
- Validated end-to-end on a real Qwen3 dense triple (non-streaming + streaming + grammar-disable paths).

**Deliberately deferred (not in v1) — file follow-ups if wanted:**
- Per-model \`models.json\` overrides for the three \`speculative_proxy_*\` fields (v1 is global/env-only; strategy itself is per-model selectable).
- Hybrid / GDN families (Qwen3.5/Next, Gemma 3) — blocked by the single-GDN-class \`GDNStateCapture\` constraint; v1 is dense-only.
- "Grammar applied *after* combination" — v1 disables proxy-tuning per-request under grammar, like every other speculative strategy.
